pennylane/wires.py

```diff
@@ -0,0 +1,322 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :class:`Wires` class, which takes care of wire bookkeeping.
+"""
+from collections import Sequence, Iterable
+import numpy as np  # for random functions
+from copy import deepcopy
+
+# tolerance for integer-like wire indices
+TOLERANCE = 1e-8
+
+
+class WireError(Exception):
+    """Exception raised by a :class:`~.pennylane.wires.Wire` object when it is unable to process wires.
+    """
+
+
+def _clean(iterable):
+    """Converts 'iterable' into list of integers, and checks that all integers are unique and non-negative.
+
+        Args:
+            iterable (Iterable): some iterable object that represents a sequence of wire indices
+
+        Returns:
+            list[int]: cleaned wire index sequence
+
+        Raises:
+            WireError if elements cannot be turned into integers, if they are negative or not unique.
+        """
+
+    iterable = list(iterable)
+
+    # Turn elements into integers, if possible
+    for idx, w in enumerate(iterable):
+
+        if not isinstance(w, int):
+            # Try to parse integer-like elements to int
+            try:
+                # This works for floats and numpy.int
+                difference = abs(w - round(w))
+                if difference < TOLERANCE:
+                    iterable[idx] = int(w)
+                else:
+                    raise TypeError
+            except TypeError:
+                raise WireError(
+                    "Wire indices must be integers; got {} of type {}.".format(w, type(w))
+                )
```

Comment:
Depending how often this gets called, it might be worth rewriting this using NumPy broadcasting rather than loops. However then there is the added overhead of creating a numpy array.

Reply:
I see from further below that this is only called on __init__ :)

Reply:
Yes, and we can optimize it once there are devices with more than 10 wires on average :P